ref(docker): create a standalone docker artifact to run `lightwalletd`

gustavovalverde opened the pull request:
Please ensure this checklist is followed for any pull requests for this repo. This checklist must be checked by both the PR creator and by anyone who reviews the PR.
* [ ] Relevant documentation for this PR has to be completed before the PR can be merged
* [ ] A test plan for the PR must be documented in the PR notes and included in the test plan for the next regular release

As a note, all CI tests need to be passing and all appropriate code reviews need to be done before this PR can be merged

----

## Motivation

This PR is meant to align lightwalletd docker execution to what we previously had working in https://github.com/ZcashFoundation/zebra/pull/8323 (but we no longer need in our repo). Allowing users to use an standalone lightwalletd docker image without further configurations, allowing users to just do:

```shell
docker run --detach -it -p 9067:9067 -p 9068:9068 electriccoinco/lightwalletd
```

And have a working lightwalletd server out-of-the-box (at least for testing/development purposes)

### Changes
- Upgrade to `golang:1.19` as https://github.com/zcash/lightwalletd/commit/1002f42497c17693fae1a09aafba6b68b2e640b1 made this a requirement, considering some dependencies no longer work with `golang:1.17`
- Use `go` instead of `make` to build inside the image (this is the expected approach with Docker)
- Remove instructions with `docker_img_run_zcashd` as we don't have `zcashd nor `zcash-cli` inside the container
- Add a `.dockerignore` file to just have the required file inside the container
- Use a multistage Dockerfile to have a resulting image with the bare minimum: `lightwalletd`, and with a considerably reduced image size
- Set some variables as `ARGs`, so people can change their values if they'd like to build the image locally. 
- Create a default `zcash.conf` to use this image as a standalone artifact without further configurations
- Use the available `cert.key` and `cert.pem` from the repo to avoid running it with `--no-tls-very-insecure` by default

### Testing

You can test this using:
```shell
docker build --target runtime --tag local/lightwalletd .
docker run -p 9067:9067 -p 9068:9068  local/lightwalletd
```

If you'd like to test this running alongside `Zebra`, you can test with:
```shell
docker-compose -f docker-compose.zebra.yml up --build -d
```

gustavovalverde committed:
ref(docker): create a standalone Dockerfile to run `lightwalletd`

gustavovalverde commented:
Also, as this is now using a multistage Dockerfile, this can be easily extended to add a `tests` stage in it, to run with CI

gustavovalverde committed:
Merge branch 'master' into ref-dockerfile

daira commented on Dockerfile:
```diff
@@ -1,22 +1,110 @@
-FROM golang:1.22 AS lightwalletd_base
+# /************************************************************************
+ #  File: Dockerfile
+ #  Author: gustavovalverde
```
```suggestion
 #  Author: gustavovalverde, mdr0id
```

daira reviewed (dismissed):
utACK with suggestion.

gustavovalverde committed:
chore: reduce diff after conflict

gustavovalverde commented:
Thank you @daira! I actually ended up removing the whole section with comments, as this was recently removed anyways https://github.com/zcash/lightwalletd/commit/38b751e5513dd998cfda2a691800e8ed092f09c1

gustavovalverde commented on Makefile:
```diff
@@ -102,31 +102,33 @@ lwd-api.html: walletrpc/compact_formats.proto walletrpc/service.proto
 # Remove and delete ALL images and containers in Docker; assumes containers are stopped
-docker_remove_all:
-	docker system prune -f
+# docker_remove_all:
+# 	docker system prune -f
```
Solved based on this comment: https://github.com/zcash/lightwalletd/pull/476#issuecomment-1997798467

gustavovalverde commented on Makefile:
```diff
@@ -102,31 +102,33 @@ lwd-api.html: walletrpc/compact_formats.proto walletrpc/service.proto
 docker_img_run_lightwalletd_insecure_server:
 	docker exec -i zcashdlwd server --no-tls-very-insecure=true --conf-file /home/zcash/.zcash/zcash.conf --log-file /logs/server.log --bind-addr 127.0.0.1:18232
 
+# TODO: remove this command as its destructive for people using Docker for other purposes
 # Remove and delete ALL images and containers in Docker; assumes containers are stopped
```
Solved based on this comment: https://github.com/zcash/lightwalletd/pull/476#issuecomment-1997798467

LarryRuane commented:
I don't understand how to run this (to test it). I'm sure part of the problem is that I'm unfamiliar with docker. For example, am I expected to start a `zcashd` instance manually, or does this happen within docker? If I start it manually, should I give any non-default command line (or `zcash.conf`) options? Should I be using the `make` targets, and if so, which ones? Can you document the exact steps need to run this? Actually, I think updating `docs/docker-run.md` might be a good place to do this, so people don't have to refer to this PR.

Also I see that this PR needs to be rebased onto the latest master branch. (I did that here locally, it wasn't difficult, but if you'd like I can push my branch to GitHub.)

mpguerra commented:
I see this has been approved now, thank you! Is there anything in particular that is preventing this PR from merging?

gustavovalverde commented:
@LarryRuane I'll be adding the process and documentation so it's clearer and easier to test this changes.

LarryRuane commented:
@gustavovalverde 
> Is there anything in particular that is preventing this PR from merging?

No, we can merge it, but I just wanted to do some testing myself as an extra safety margin. But since this PR makes only docker-related changes, it's safe for anyone not using docker.

Now that @gustavovalverde will be adding some documentation, we could wait for this and have me test it. Or would you rather have me merge it right away? I'm okay with either plan, I don't want to hold anyone up.

gustavovalverde committed:
ref(docker): add a `docker-compose` to run lighwalletd with zebra

This also includes documentation on how to run it

gustavovalverde commented:
Hi @LarryRuane!

I've added extra testing steps into the PR body, and also in the documentation.

I added an extra `docker-compose` file to keep the changes as minimum as possible.

gustavovalverde committed:
Merge branch 'master' into ref-dockerfile
